enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,172 @@
+
+4.  The CLI in-cluster manager can provide CLIs as plugins through [krew](https://github.com/kubernetes-sigs/krew).  CLI images will serve the artifacts to enable disconnected downloads, and OpenShift CLIs will be available as 
+[krew plugins](https://github.com/kubernetes-sigs/krew-index/tree/master/plugins).
+    * Krew and Krew plugins are upstream projects that Kubernetes users are already familiar with
+    * From the [README](https://github.com/kubernetes-sigs/krew/#what-does-krew-do): `Krew is a tool that makes it easy to use kubectl plugins. Krew helps you discover plugins, install and manage them on your machine. It is similar to tools like apt, dnf or brew. Today, over 70 kubectl plugins are available on Krew.` It makes sense to offer OpenShift CLIs in the same way.
+    * With krew we will package artifacts within an image that will populate an index.  Mirrored local registries will provide binaries to disconnected environments.  
+
+Option 4 is the option we as a team are leaning toward.  It requires that each OpenShift CLI publish artifacts, for example, [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/)
+then create an image similar to this: [example cli-artifacts image](https://github.com/openshift/oc/blob/master/images/cli-artifacts/Dockerfile.rhel).  CLI artifacts are packaged in images.
+The CLI in-cluster manager will gather information from each CLI's CR about each CLI and its image.  A Krew index will be populated in every cluster to provide available CLIs and plugins for a cluster.
+
+As laid out in this proposal, each CLI will provide a separate image.
+Each CLI is responsible for creating a CR to hold metadata that `oc krew` will depend on.  The CR will serve to deliver the metadata and description
+of its deliverable binary.  An in-cluster CLI manager controller will access the CRs to populate and maintain an in-cluster Krew Index.  Users will install OpenShift tools that are known compatible
+with each cluster version through `oc krew`.    
+
+A CLI must provide a ClusterCLI CR.  The result of this proposal will be:
+* ClusterCLI Custom Resource Definition
+* Krew Index and a controller to create the index from CLI CRs.
+* mechanism to manage CLIs via ClusterCLI CRs and Krew
+* a reference consumer implementation, oc, that other CLIs can follow
+
+### Requirements:
+
+1. CLI author requirements:
+    - provide CLI image containing binaries
+    - Custom Resource that describes the binaries (location within the image, metadata, description)
+2. An operator that will be responsible for:
+    - registering the new CLI CRD 
+    - reading metadata from CR from authors and extracting the binaries from the CLI images to local disk
+    - serving extracted binaries
+    - serving krew-compatible index
+3. oc changes:
+    - embed krew
+
+
+### Non-Goals
```

Comment:
Non-goals should right after goals, pls move.

Reply:
done